### ISSUE TITLE
BN inserts seemingly arbitrary padding in `__packed` structure

### ISSUE THREAD

**Issue by alexrp** [Vector35/binaryninja-api#4502]
**Version and Platform (required):**
- Binary Ninja Version: 3.5.4402-dev Personal (c2f291a7)
- OS: Windows
- OS Version: 10.0.25375.1
- CPU Architecture: x64

**Bug Description:**
I have this structure:

```c
struct __base(IS1MemoryObject, 0) S1GameObject __packed
{
    struct `S1GameObject::__vftable`* __vftable;
    BOOL field_8;
    BOOL field_c;
    int32_t name_plate_height;
    int32_t field_14;
    int32_t field_18;
    enum S1GameObjectFlags flags;
    BOOL initialized;
    S1GameId game_id;
    void* field_30;
    int32_t field_38;
    int32_t field_3c;
    struct S1ControllerBase* controllers[0x164];
    void* field_b60;
    int32_t field_b68;
    int32_t field_b6c;
    __padding char _B70[8];
    BOOL field_b78;
    __padding char _B7C[4];
    BOOL field_b80;
    struct FCriticalSection field_b84;
    struct `TArray<S1ControllerDescriptor>` controller_descriptors;
    void* field_bdc;
    int32_t field_be4;
    int32_t field_be8;
    int32_t field_bec;
    BOOL field_bf0;
    BOOL field_bf4;
    struct FCriticalSection field_bf8;
    int64_t unique_id;
    S1GameId field_c48;
};

```

If I change it to have all the elements of `controllers` laid out as individual fields, BN inserts unwanted padding in the structure.

**Steps To Reproduce:**
1. Grab [this BNDB](https://drive.google.com/file/d/1m_BcdR_CVfYNsTZ_YMIPFxWazXBO-ymI/view?usp=sharing).
2. Go to Types and find `S1GameObject`.
3. Redefine it as follows:

<details>
<p>

```c
struct __base(IS1MemoryObject, 0) S1GameObject __packed
{
    struct `S1GameObject::__vftable`* __vftable;
    BOOL field_8;
    BOOL field_c;
    int32_t name_plate_height;
    int32_t field_14;
    int32_t field_18;
    enum S1GameObjectFlags flags;
    BOOL initialized;
    S1GameId game_id;
    void* field_30;
    int32_t field_38;
    int32_t field_3c;
    struct S1ControllerBase* field_40;
    struct S1ControllerBase* field_48;
    struct S1ControllerBase* field_50;
    struct S1ControllerBase* field_58;
    struct S1ControllerBase* field_60;
    struct S1ControllerBase* field_68;
    struct S1ControllerBase* field_70;
    struct S1ControllerBase* field_78;
    struct S1ControllerBase* field_80;
    struct S1ControllerBase* field_88;
    struct S1ControllerBase* field_90;
    struct S1ControllerBase* field_98;
    struct S1ControllerBase* field_a0;
    struct S1ControllerBase* field_a8;
    struct S1ControllerBase* field_b0;
    struct S1ControllerBase* field_b8;
    struct S1ControllerBase* field_c0;
    struct S1ControllerBase* field_c8;
    struct S1ControllerBase* field_d0;
    struct S1ControllerBase* field_d8;
    struct S1ControllerBase* field_e0;
    struct S1ControllerBase* field_e8;
    struct S1ControllerBase* field_f0;
    struct S1ControllerBase* field_f8;
    struct S1ControllerBase* field_100;
    struct S1ControllerBase* field_108;
    struct S1ControllerBase* field_110;
    struct S1ControllerBase* field_118;
    struct S1ControllerBase* field_120;
    struct S1ControllerBase* field_128;
    struct S1ControllerBase* field_130;
    struct S1ControllerBase* field_138;
    struct S1ControllerBase* field_140;
    struct S1ControllerBase* field_148;
    struct S1ControllerBase* field_150;
    struct S1ControllerBase* field_158;
    struct S1ControllerBase* field_160;
    struct S1ControllerBase* field_168;
    struct S1ControllerBase* field_170;
    struct S1ControllerBase* field_178;
    struct S1ControllerBase* field_180;
    struct S1ControllerBase* field_188;
    struct S1ControllerBase* field_190;
    struct S1ControllerBase* field_198;
    struct S1ControllerBase* field_1a0;
    struct S1ControllerBase* field_1a8;
    struct S1ControllerBase* field_1b0;
    struct S1ControllerBase* field_1b8;
    struct S1ControllerBase* field_1c0;
    struct S1ControllerBase* field_1c8;
    struct S1ControllerBase* field_1d0;
    struct S1ControllerBase* field_1d8;
    struct S1ControllerBase* field_1e0;
    struct S1ControllerBase* field_1e8;
    struct S1ControllerBase* field_1f0;
    struct S1ControllerBase* field_1f8;
    struct S1ControllerBase* field_200;
    struct S1ControllerBase* field_208;
    struct S1ControllerBase* field_210;
    struct S1ControllerBase* field_218;
    struct S1ControllerBase* field_220;
    struct S1ControllerBase* field_228;
    struct S1ControllerBase* field_230;
    struct S1ControllerBase* field_238;
    struct S1ControllerBase* field_240;
    struct S1ControllerBase* field_248;
    struct S1ControllerBase* field_250;
    struct S1ControllerBase* field_258;
    struct S1ControllerBase* field_260;
    struct S1ControllerBase* field_268;
    struct S1ControllerBase* field_270;
    struct S1ControllerBase* field_278;
    struct S1ControllerBase* field_280;
    struct S1ControllerBase* field_288;
    struct S1ControllerBase* field_290;
    struct S1ControllerBase* field_298;
    struct S1ControllerBase* field_2a0;
    struct S1ControllerBase* field_2a8;
    struct S1ControllerBase* field_2b0;
    struct S1ControllerBase* field_2b8;
    struct S1ControllerBase* field_2c0;
    struct S1ControllerBase* field_2c8;
    struct S1ControllerBase* field_2d0;
    struct S1ControllerBase* field_2d8;
    struct S1ControllerBase* field_2e0;
    struct S1ControllerBase* field_2e8;
    struct S1ControllerBase* field_2f0;
    struct S1ControllerBase* field_2f8;
    struct S1ControllerBase* field_300;
    struct S1ControllerBase* field_308;
    struct S1ControllerBase* field_310;
    struct S1ControllerBase* field_318;
    struct S1ControllerBase* field_320;
    struct S1ControllerBase* field_328;
    struct S1ControllerBase* field_330;
    struct S1ControllerBase* field_338;
    struct S1ControllerBase* field_340;
    struct S1ControllerBase* field_348;
    struct S1ControllerBase* field_350;
    struct S1ControllerBase* field_358;
    struct S1ControllerBase* field_360;
    struct S1ControllerBase* field_368;
    struct S1ControllerBase* field_370;
    struct S1ControllerBase* field_378;
    struct S1ControllerBase* field_380;
    struct S1ControllerBase* field_388;
    struct S1ControllerBase* field_390;
    struct S1ControllerBase* field_398;
    struct S1ControllerBase* field_3a0;
    struct S1ControllerBase* field_3a8;
    struct S1ControllerBase* field_3b0;
    struct S1ControllerBase* field_3b8;
    struct S1ControllerBase* field_3c0;
    struct S1ControllerBase* field_3c8;
    struct S1ControllerBase* field_3d0;
    struct S1ControllerBase* field_3d8;
    struct S1ControllerBase* field_3e0;
    struct S1ControllerBase* field_3e8;
    struct S1ControllerBase* field_3f0;
    struct S1ControllerBase* field_3f8;
    struct S1ControllerBase* field_400;
    struct S1ControllerBase* field_408;
    struct S1ControllerBase* field_410;
    struct S1ControllerBase* field_418;
    struct S1ControllerBase* field_420;
    struct S1ControllerBase* field_428;
    struct S1ControllerBase* field_430;
    struct S1ControllerBase* field_438;
    struct S1ControllerBase* field_440;
    struct S1ControllerBase* field_448;
    struct S1ControllerBase* field_450;
    struct S1ControllerBase* field_458;
    struct S1ControllerBase* field_460;
    struct S1ControllerBase* field_468;
    struct S1ControllerBase* field_470;
    struct S1ControllerBase* field_478;
    struct S1ControllerBase* field_480;
    struct S1ControllerBase* field_488;
    struct S1ControllerBase* field_490;
    struct S1ControllerBase* field_498;
    struct S1ControllerBase* field_4a0;
    struct S1ControllerBase* field_4a8;
    struct S1ControllerBase* field_4b0;
    struct S1ControllerBase* field_4b8;
    struct S1ControllerBase* field_4c0;
    struct S1ControllerBase* field_4c8;
    struct S1ControllerBase* field_4d0;
    struct S1ControllerBase* field_4d8;
    struct S1ControllerBase* field_4e0;
    struct S1ControllerBase* field_4e8;
    struct S1ControllerBase* field_4f0;
    struct S1ControllerBase* field_4f8;
    struct S1ControllerBase* field_500;
    struct S1ControllerBase* field_508;
    struct S1ControllerBase* field_510;
    struct S1ControllerBase* field_518;
    struct S1ControllerBase* field_520;
    struct S1ControllerBase* field_528;
    struct S1ControllerBase* field_530;
    struct S1ControllerBase* field_538;
    struct S1ControllerBase* field_540;
    struct S1ControllerBase* field_548;
    struct S1ControllerBase* field_550;
    struct S1ControllerBase* field_558;
    struct S1ControllerBase* field_560;
    struct S1ControllerBase* field_568;
    struct S1ControllerBase* field_570;
    struct S1ControllerBase* field_578;
    struct S1ControllerBase* field_580;
    struct S1ControllerBase* field_588;
    struct S1ControllerBase* field_590;
    struct S1ControllerBase* field_598;
    struct S1ControllerBase* field_5a0;
    struct S1ControllerBase* field_5a8;
    struct S1ControllerBase* field_5b0;
    struct S1ControllerBase* field_5b8;
    struct S1ControllerBase* field_5c0;
    struct S1ControllerBase* field_5c8;
    struct S1ControllerBase* field_5d0;
    struct S1ControllerBase* field_5d8;
    struct S1ControllerBase* field_5e0;
    struct S1ControllerBase* field_5e8;
    struct S1ControllerBase* field_5f0;
    struct S1ControllerBase* field_5f8;
    struct S1ControllerBase* field_600;
    struct S1ControllerBase* field_608;
    struct S1ControllerBase* field_610;
    struct S1ControllerBase* field_618;
    struct S1ControllerBase* field_620;
    struct S1ControllerBase* field_628;
    struct S1ControllerBase* field_630;
    struct S1ControllerBase* field_638;
    struct S1ControllerBase* field_640;
    struct S1ControllerBase* field_648;
    struct S1ControllerBase* field_650;
    struct S1ControllerBase* field_658;
    struct S1ControllerBase* field_660;
    struct S1ControllerBase* field_668;
    struct S1ControllerBase* field_670;
    struct S1ControllerBase* field_678;
    struct S1ControllerBase* field_680;
    struct S1ControllerBase* field_688;
    struct S1ControllerBase* field_690;
    struct S1ControllerBase* field_698;
    struct S1ControllerBase* field_6a0;
    struct S1ControllerBase* field_6a8;
    struct S1ControllerBase* field_6b0;
    struct S1ControllerBase* field_6b8;
    struct S1ControllerBase* field_6c0;
    struct S1ControllerBase* field_6c8;
    struct S1ControllerBase* field_6d0;
    struct S1ControllerBase* field_6d8;
    struct S1ControllerBase* field_6e0;
    struct S1ControllerBase* field_6e8;
    struct S1ControllerBase* field_6f0;
    struct S1ControllerBase* field_6f8;
    struct S1ControllerBase* field_700;
    struct S1ControllerBase* field_708;
    struct S1ControllerBase* field_710;
    struct S1ControllerBase* field_718;
    struct S1ControllerBase* field_720;
    struct S1ControllerBase* field_728;
    struct S1ControllerBase* field_730;
    struct S1ControllerBase* field_738;
    struct S1ControllerBase* field_740;
    struct S1ControllerBase* field_748;
    struct S1ControllerBase* field_750;
    struct S1ControllerBase* field_758;
    struct S1ControllerBase* field_760;
    struct S1ControllerBase* field_768;
    struct S1ControllerBase* field_770;
    struct S1ControllerBase* field_778;
    struct S1ControllerBase* field_780;
    struct S1ControllerBase* field_788;
    struct S1ControllerBase* field_790;
    struct S1ControllerBase* field_798;
    struct S1ControllerBase* field_7a0;
    struct S1ControllerBase* field_7a8;
    struct S1ControllerBase* field_7b0;
    struct S1ControllerBase* field_7b8;
    struct S1ControllerBase* field_7c0;
    struct S1ControllerBase* field_7c8;
    struct S1ControllerBase* field_7d0;
    struct S1ControllerBase* field_7d8;
    struct S1ControllerBase* field_7e0;
    struct S1ControllerBase* field_7e8;
    struct S1ControllerBase* field_7f0;
    struct S1ControllerBase* field_7f8;
    struct S1ControllerBase* field_800;
    struct S1ControllerBase* field_808;
    struct S1ControllerBase* field_810;
    struct S1ControllerBase* field_818;
    struct S1ControllerBase* field_820;
    struct S1ControllerBase* field_828;
    struct S1ControllerBase* field_830;
    struct S1ControllerBase* field_838;
    struct S1ControllerBase* field_840;
    struct S1ControllerBase* field_848;
    struct S1ControllerBase* field_850;
    struct S1ControllerBase* field_858;
    struct S1ControllerBase* field_860;
    struct S1ControllerBase* field_868;
    struct S1ControllerBase* field_870;
    struct S1ControllerBase* field_878;
    struct S1ControllerBase* field_880;
    struct S1ControllerBase* field_888;
    struct S1ControllerBase* field_890;
    struct S1ControllerBase* field_898;
    struct S1ControllerBase* field_8a0;
    struct S1ControllerBase* field_8a8;
    struct S1ControllerBase* field_8b0;
    struct S1ControllerBase* field_8b8;
    struct S1ControllerBase* field_8c0;
    struct S1ControllerBase* field_8c8;
    struct S1ControllerBase* field_8d0;
    struct S1ControllerBase* field_8d8;
    struct S1ControllerBase* field_8e0;
    struct S1ControllerBase* field_8e8;
    struct S1ControllerBase* field_8f0;
    struct S1ControllerBase* field_8f8;
    struct S1ControllerBase* field_900;
    struct S1ControllerBase* field_908;
    struct S1ControllerBase* field_910;
    struct S1ControllerBase* field_918;
    struct S1ControllerBase* field_920;
    struct S1ControllerBase* field_928;
    struct S1ControllerBase* field_930;
    struct S1ControllerBase* field_938;
    struct S1ControllerBase* field_940;
    struct S1ControllerBase* field_948;
    struct S1ControllerBase* field_950;
    struct S1ControllerBase* field_958;
    struct S1ControllerBase* field_960;
    struct S1ControllerBase* field_968;
    struct S1ControllerBase* field_970;
    struct S1ControllerBase* field_978;
    struct S1ControllerBase* field_980;
    struct S1ControllerBase* field_988;
    struct S1ControllerBase* field_990;
    struct S1ControllerBase* field_998;
    struct S1ControllerBase* field_9a0;
    struct S1ControllerBase* field_9a8;
    struct S1ControllerBase* field_9b0;
    struct S1ControllerBase* field_9b8;
    struct S1ControllerBase* field_9c0;
    struct S1ControllerBase* field_9c8;
    struct S1ControllerBase* field_9d0;
    struct S1ControllerBase* field_9d8;
    struct S1ControllerBase* field_9e0;
    struct S1ControllerBase* field_9e8;
    struct S1ControllerBase* field_9f0;
    struct S1ControllerBase* field_9f8;
    struct S1ControllerBase* field_a00;
    struct S1ControllerBase* field_a08;
    struct S1ControllerBase* field_a10;
    struct S1ControllerBase* field_a18;
    struct S1ControllerBase* field_a20;
    struct S1ControllerBase* field_a28;
    struct S1ControllerBase* field_a30;
    struct S1ControllerBase* field_a38;
    struct S1ControllerBase* field_a40;
    struct S1ControllerBase* field_a48;
    struct S1ControllerBase* field_a50;
    struct S1ControllerBase* field_a58;
    struct S1ControllerBase* field_a60;
    struct S1ControllerBase* field_a68;
    struct S1ControllerBase* field_a70;
    struct S1ControllerBase* field_a78;
    struct S1ControllerBase* field_a80;
    struct S1ControllerBase* field_a88;
    struct S1ControllerBase* field_a90;
    struct S1ControllerBase* field_a98;
    struct S1ControllerBase* field_aa0;
    struct S1ControllerBase* field_aa8;
    struct S1ControllerBase* field_ab0;
    struct S1ControllerBase* field_ab8;
    struct S1ControllerBase* field_ac0;
    struct S1ControllerBase* field_ac8;
    struct S1ControllerBase* field_ad0;
    struct S1ControllerBase* field_ad8;
    struct S1ControllerBase* field_ae0;
    struct S1ControllerBase* field_ae8;
    struct S1ControllerBase* field_af0;
    struct S1ControllerBase* field_af8;
    struct S1ControllerBase* field_b00;
    struct S1ControllerBase* field_b08;
    struct S1ControllerBase* field_b10;
    struct S1ControllerBase* field_b18;
    struct S1ControllerBase* field_b20;
    struct S1ControllerBase* field_b28;
    struct S1ControllerBase* field_b30;
    struct S1ControllerBase* field_b38;
    struct S1ControllerBase* field_b40;
    struct S1ControllerBase* field_b48;
    struct S1ControllerBase* field_b50;
    struct S1ControllerBase* field_b58;
    void* field_b60;
    int32_t field_b68;
    int32_t field_b6c;
    int64_t field_b70;
    BOOL field_b78;
    int32_t field_b7c;
    BOOL field_b80;
    struct FCriticalSection field_b84;
    struct `TArray<S1ControllerDescriptor>` controller_descriptors;
    void* field_bdc;
    int32_t field_be4;
    int32_t field_be8;
    int32_t field_bec;
    BOOL field_bf0;
    BOOL field_bf4;
    struct FCriticalSection field_bf8;
    int64_t unique_id;
    S1GameId field_c48;
};
```

</p>
</details>

4. Observe that BN inserts unwanted padding in the middle and at the end of the structure, breaking its layout and size.

![image](https://github.com/Vector35/binaryninja-api/assets/44076/2f7cda3d-2002-4be0-bb1b-9eaf6d51f9f7)
![image](https://github.com/Vector35/binaryninja-api/assets/44076/7e9455b1-7222-4e53-a7fc-cd339f2c041d)

**Expected Behavior:**
No unwanted padding.

**Comment by alexrp** [comment]
Hoping there's a workaround for this as representing `controllers` as an array makes analysis much harder; I can't give a more specific type to each element. 🙁

**Comment by CouleeApps** [comment]
Looks like the alignment on the flags are pushing it down to the next multiple of 8 bytes. Also, trying to use alignas(4) on the enum throws a clang error: `error: input.hpp: 9:5 requested alignment is less than minimum alignment of 8 for type 'enum S1GameObjectFlags'`. I'll look and see if there's an easy way around this, but it might involve another patch to clang.

**Comment by CouleeApps** [comment]
Looks like this is an issue with explicit alignment interacting with the packed extensions. Binja inserts explicit alignment attributes on types passed to clang so that clang's record layout builder can align them the same, but when packed structures are used, the explicit alignment is supposed to win over the packed.

Effectively, binja is doing the following:

    enum __attribute__((aligned(8))) S1GameObjectFlags {};
    struct __attribute__((packed)) test {
        int a;
        S1GameObjectFlags b;
    };

The aligned(8) is taking precedence over the packed, and therefore `b` is getting moved to 8 bytes in. Interestingly, this only happens for windows targets. for linux/macos, clang instead respects the packed attribute and aligns to 4 bytes. I think this might be a bug in clang, actually. Everywhere else honors the packed attribute over any alignment attributes. Looking at the clang source, it's probably this:

https://github.com/llvm/llvm-project/blob/ba877dc57afa729195519f5c2e9f37be4f38edef/clang/lib/AST/RecordLayoutBuilder.cpp#L2711-L2716

Specifically line 2716 being at the end. I'll try patching clang to reorder those lines and see if it fixes this.

**Comment by CouleeApps** [comment]
Yep, clang patch fixed this (and didn't break any of clang's tests, which makes me suspicious that they didn't test this at all). Fix will be going out Soon (office AC is currently broken so no builds going up right now, especially not with a new clang).

**Comment by alexrp** [comment]
@CouleeApps just checking, did the fix for this one land in dev builds yet? :eyes:

**Comment by CouleeApps** [comment]
In fact it did not. I looked closer at what clang was doing and got a bit lost in the sauce testing what `__attribute__` even means for MSVC mode... I'll pick it back up tomorrow though

**Comment by CouleeApps** [comment]
So further progress on this: `__packed` does the same thing in linux/macos mode as the undesired windows behavior. I'm starting to think maybe the way BN reports alignment on structures is using the wrong facilities in clang and it would be better to revise that than to try and patch the structure layout code. Needs more investigation though, so it's not as simple of a fix as I expected.

**Comment by CouleeApps** [comment]
It seems that using `#pragma pack (push, 1)` and `#pragma pack (pop)` around the definition of the type is a workaround that convinces clang to actually pack the members properly. I'm looking into replicating this through a custom attribute (it seems like the pragma uses `MaxFieldAlignmentAttr` internally), but until now you should be able to include those pragmas as a workaround.

**Comment by CouleeApps** [comment]
After poking this more on a Saturday than I'd like to admit, I think I have arrived at a solution. The `pragma pack` solution as mentioned previously probably will not work, but only on Windows. Clang uses a different structure layout algorithm for MS-ABI which considers type alignment (via `alignas()` which binja uses) stronger than any sort of `pragma pack` or `__attribute__((packed))`. Other platforms use a different algorithm and are not affected. I don't want to break MS-ABI compatibility when loading headers like windows.h, but also I want binja's `__packed` attribute to be the same on all platforms, which means it needs to be stronger on windows. So to solve both of these situations, I'm adding an extra compiler option to our build of clang that enables this stronger packing behavior, and I'm having it enable that option when the binja "use builtin macros" type parser option is enabled, which affects the type editor but not importing headers (when disabled, which is the default). Testing this was quite a mess! Hopefully this resolves the alignment issue in all of the cases you expect, though!

**Comment by alexrp** [comment]
This turned into quite the rabbit hole. 🤔 Thanks for the effort!

**Comment by CouleeApps** [comment]
This has been pushed and will be fixed in dev 3.5.4428
